lib/spack/spack/test/test_activations.py

```diff
@@ -78,7 +43,23 @@ def create_dir_structure(tmpdir, dir_structure):
 
 
 @pytest.fixture()
-def python_and_extension_dirs(tmpdir):
+def builtin_and_mock_packages():
+    # These tests use mock_repo packages to test functionality of builtin
+    # packages. They therefore put the mock repo UNDER the builtin repo.
+    old_path = spack.repo.path
+    repo_dirs = [spack.paths.packages_path, spack.paths.mock_packages_path]
+    path = RepoPath(*repo_dirs)
+    sys.meta_path.append(path)
```

Comment:
Could this use `repo.swap`. Although there is minimal duplication, it's useful to minimize the number of unique manipulations of `sys.meta_path`.

Reply:
I'm a little bit worried that `repo.swap` doesn't restore `sys.meta_path` properly, but that's a separate scope of work than this PR and this PR should use the logic there. We can fix it separately if there's an issue.

If the path added by `repo.swap` is already in `sys.meta_path`, won't `repo.swap` accidentally change it from its current location to the end of `sys.meta_path` (because `list.remove` removes the first instance in the list, not the last)? If so, we can fix this by either caching `sys.meta_path` the same way we cache `repo.path` or by reversing the list, removing the item, and reversing it again.

Reply:
> If the path added by repo.swap is already in sys.meta_path, won't repo.swap accidentally change it from its current location to the end of sys.meta_path (because list.remove removes the first instance in the list, not the last)?

I think you're right. Generally I think this just hasn't happened yet - the same `Repo`/`RepoPath` is not added multiple times. I think it's worth sorting out in a separate PR. Generally I assume that it should be disallowed to add it multiple times (raise an exception).

Reply:
Testing could be a reason to add the same one multiple times, but if we need that there's other ways to ensure we don't mess up the ordering.